### PR TITLE
feat(ENG-1783)!:  kill environments

### PR DESCRIPTION
**Pull request by itayd**
This is a breaking change that kills the environment objects. now everything belongs to projects.

TODO:
- [x] fix tests
- [x] do proper migration

### CODE REVIEW

File: go.sum
Comment (daabr) [inline go.sum:1]: New entries in `go.sum` without `go.mod`? Does this file belong in this PR / is the other file missing by accident?

File: cmd/ak/cmd/deploy.go
Comment (daabr) [inline cmd/ak/cmd/deploy.go:68]: Remove the `env` var at the top of this file, and the `env` flag at the bottom.

File: cmd/ak/cmd/deploy.go
Comment (daabr) [inline cmd/ak/cmd/deploy.go:81]: Question: is this really needed? Build objects already have a project ID field.

File: cmd/ak/cmd/deployments/create.go
Comment (daabr) [inline cmd/ak/cmd/deployments/create.go:17]: Same question as in `cmd/deploy.go`: if build objects already have a project ID field, why does the user need to specify it manually here?

File: proto/autokitteh/store/v1/svc.proto
Comment (daabr) [inline proto/autokitteh/store/v1/svc.proto:14]: I'm fine with this, but to be clear you're committing multiple violations here at a point in time that AK is no longer in its "early days":
- Delete a field instead of deprecating it
- Change a field ID

**Review by daabr — COMMENTED** [review]
I really liked the concept of envs, but this is the right thing to do for now, so yay code cleanup! Left a few small comments - will approve once you reply to / address them, and fix all the tests and lint warnings.

File: go.sum
Comment (itayd) [reply go.sum:1]: /shrug
i just ran go tidy again

File: cmd/ak/cmd/deploy.go
Comment (itayd) [reply cmd/ak/cmd/deploy.go:68]: done

File: proto/autokitteh/store/v1/svc.proto
Comment (itayd) [reply proto/autokitteh/store/v1/svc.proto:14]: nobody is using the old proto yet, so I'm fine doing these kind of stuff while we can get away with it.

File: cmd/ak/cmd/deploy.go
Comment (itayd) [reply cmd/ak/cmd/deploy.go:81]: i think having project id in deployment is important, having a project_id in build is only for auth. I'm ok with leaving this as is for now. we might revisit in the future.

File: cmd/ak/cmd/deployments/create.go
Comment (itayd) [reply cmd/ak/cmd/deployments/create.go:17]: see above

**Comment by efiShtain** [comment]
Just noting this requires a lot of UI updates I guess, need to make sure UI has time to adhere to this change before we merge

File: cmd/ak/common/build.go
Comment (efiShtain) [inline cmd/ak/common/build.go]: why do you need to return the extra argument here ?

File: cmd/ak/cmd/projects/deploy.go
Comment (efiShtain) [inline cmd/ak/cmd/projects/deploy.go:50]: the command's Use description says deploye <project name or ID>
why do we get the pid from the buildProject result instead of user input?
if this is the only reason to change buildProject signature to return projectId you can revert the change of buildProject

File: cmd/ak/cmd/triggers/create.go
Comment (efiShtain) [inline cmd/ak/cmd/triggers/create.go]:
```suggestion
		// Project is required.
```

File: cmd/ak/cmd/deploy.go
Comment (efiShtain) [inline cmd/ak/cmd/deploy.go]: again projectId is provided from user, no need to get it as result from buildProject

File: cmd/ak/cmd/deploy.go
Comment (efiShtain) [reply cmd/ak/cmd/deploy.go:81]: i also prefer leaving the projectID in deployment, it has much more use on the API and UI side
we can remove the projectID from build, since build belongs to one deployment only, we don't share builds

File: internal/backend/db/dbgorm/projects.go
Comment (efiShtain) [inline internal/backend/db/dbgorm/projects.go:56]: does this list only active deployments ? or any deployment ?

File: internal/backend/db/dbgorm/projects.go
Comment (efiShtain) [inline internal/backend/db/dbgorm/projects.go:52]: so why isn't it a transaction ?

File: internal/backend/db/dbgorm/projects.go
Comment (efiShtain) [inline internal/backend/db/dbgorm/projects.go]: deployment now has project id set, no need to join projects 
I'm also not sure whay the signature of this function is getProjectDeployments while is returns a list of states

this can be simplified to 
```sql 
select deployment_id, state from deployments where project_id = pid
```

no join and no need to distinct

the function should also be renamed to getProjectDeploymentsStates or it should get a fields list to return like
```go
func (db *gormdb) getProjectDeployments(ctx context.Contest, pid sdktypes.ProjectId, ...fields string)
```

File: internal/backend/db/dbgorm/triggers.go
Comment (efiShtain) [inline internal/backend/db/dbgorm/triggers.go:85]: this is not backward compatible if i understand correctly,
do we hav a migratino plan ?

File: internal/backend/sessions/sessiondata/data.go
Comment (efiShtain) [inline internal/backend/sessions/sessiondata/data.go:56]: why get session data download the build data ?
it's not related to this PR but i just saw it here

File: internal/backend/sessions/sessiondata/data.go
Comment (efiShtain) [inline internal/backend/sessions/sessiondata/data.go:63]: i think env wasn't required before
but can we have a session without a valid projectID ?
maybe it should return an error if the projectID is invalid on the session ?

File: internal/backend/sessions/sessionworkflows/modules/store/store.go
Comment (efiShtain) [inline internal/backend/sessions/sessionworkflows/modules/store/store.go:25]: again breaking change,
it's not only API breaking change, its data breaking change that won't be so easy to migrate

File: internal/backend/store/store.go
Comment (efiShtain) [inline internal/backend/store/store.go:79]: do we still need the prefix function ?

File: internal/manifest/plan.go
Comment (efiShtain) [inline internal/manifest/plan.go]: remove comment

File: internal/manifest/plan.go
Comment (efiShtain) [inline internal/manifest/plan.go:99]: is this something new?
i don't see this code before
why do we need it? am i missing something ?

File: migrations/postgres/20241106064610_killenv.sql
Comment (efiShtain) [inline migrations/postgres/20241106064610_killenv.sql]: did you check this migrations on a full db with actual data ?
does it require down time ?
how do we make sure running code doesn't try to create envs while you run migrations and old code still running?

Maybe we should plan this change in multiple deployments

File: proto/autokitteh/deployments/v1/svc.proto
Comment (efiShtain) [inline proto/autokitteh/deployments/v1/svc.proto:47]: all breaking API changes should be cheked with UI before merging

consider having a backward compatible change, 
mark the fields as deprecated
and if someone provided env_id still, assume it is a project_id and continue with it in the backend

once ui is ready and updated, remove the env_id

File: proto/autokitteh/dispatcher/v1/svc.proto
Comment (efiShtain) [inline proto/autokitteh/dispatcher/v1/svc.proto:29]: why do we need the project_id for re / dispatching event ? we needed the env cause it might have different arguments, but now deployment should be enough, no need for project

File: sdk/sdkservices/dispatcher.go
Comment (efiShtain) [inline sdk/sdkservices/dispatcher.go:11]: no need for project

**Review by efiShtain — COMMENTED** [review]
requires some change
a really risky change
requires a dedicated TESTED plan for migrating live environments, backups before executing the plan, might require down time

File: cmd/ak/common/build.go
Comment (itayd) [reply cmd/ak/common/build.go]: you need the project id further on, this is just kind of a shortcut.

File: internal/backend/db/dbgorm/projects.go
Comment (itayd) [reply internal/backend/db/dbgorm/projects.go:56]: looks like all deployments

Comment (itayd) [reply internal/backend/db/dbgorm/projects.go:56]: any non-deleted one that is

File: internal/backend/db/dbgorm/triggers.go
Comment (itayd) [reply internal/backend/db/dbgorm/triggers.go:85]: i'm going to re-add envID back, so we'll be able to preserve it and run some migration to move from env to proj. after migration we can remove the envid.

File: internal/backend/sessions/sessiondata/data.go
Comment (itayd) [reply internal/backend/sessions/sessiondata/data.go:56]: we need to give the runtime the build?

File: internal/backend/sessions/sessiondata/data.go
Comment (itayd) [reply internal/backend/sessions/sessiondata/data.go:63]: so for session tests there is no project id. theoretically one might launch sessions without a project.

File: internal/backend/sessions/sessionworkflows/modules/store/store.go
Comment (itayd) [reply internal/backend/sessions/sessionworkflows/modules/store/store.go:25]: nobody is using this for now, so it's ok. i might even totally remove it.

File: internal/manifest/plan.go
Comment (itayd) [reply internal/manifest/plan.go:99]: instead of setting up a default env, it's just setting all the vars. before vars belong to env, now they belong to project.

File: migrations/postgres/20241106064610_killenv.sql
Comment (itayd) [reply migrations/postgres/20241106064610_killenv.sql]: oh we're going to work on this in a separate change. we'll discuss.

File: proto/autokitteh/deployments/v1/svc.proto
Comment (itayd) [reply proto/autokitteh/deployments/v1/svc.proto:47]: we'll discuss how exactly to do this - i think we should just merge for now, but deploy in tandem backend + UI. so there wont be any crosstalk between versions. we have low enough volume rn that we can pull it off.

File: proto/autokitteh/dispatcher/v1/svc.proto
Comment (itayd) [reply proto/autokitteh/dispatcher/v1/svc.proto:29]: it's essentially an convinience that says "just throw it to the active deployment of that project".

File: sdk/sdkservices/dispatcher.go
Comment (itayd) [reply sdk/sdkservices/dispatcher.go:11]: see above

File: cmd/ak/cmd/deploy.go
Comment (itayd) [reply cmd/ak/cmd/deploy.go:81]: i'm going to leave it in as optional in build as well for now for record keeping. might remove later.

File: cmd/ak/cmd/projects/deploy.go
Comment (itayd) [reply cmd/ak/cmd/projects/deploy.go:50]: not sure i follow you here, but project in build is optional.

File: cmd/ak/cmd/deploy.go
Comment (itayd) [reply cmd/ak/cmd/deploy.go]: ok, moved some things around.

File: internal/backend/db/dbgorm/projects.go
Comment (itayd) [reply internal/backend/db/dbgorm/projects.go:52]: it is called in a transaction changed the comment.

File: internal/backend/db/dbgorm/projects.go
Comment (itayd) [reply internal/backend/db/dbgorm/projects.go]: good find. done.

File: internal/backend/store/store.go
Comment (itayd) [reply internal/backend/store/store.go:79]: i think it's better this way since it's more standard. nobody is currently using it anyway currently and we're going to redo this.

File: internal/manifest/plan.go
Comment (itayd) [reply internal/manifest/plan.go]: done

File: cmd/ak/cmd/projects/deploy.go
Comment (efiShtain) [reply cmd/ak/cmd/projects/deploy.go:50]: I say it can't be optional, has to be mandatory
you can't build anything, you have to build a project

File: cmd/ak/common/build.go
Comment (efiShtain) [reply cmd/ak/common/build.go]: I don't like that BuildProject get some random project name and returns projectID
it should get a project object or a projectID and build that

File: proto/autokitteh/dispatcher/v1/svc.proto
Comment (efiShtain) [reply proto/autokitteh/dispatcher/v1/svc.proto:29]: so why do we need the deployment_id if the only option is the active deployment any way ?

File: internal/backend/sessions/sessiondata/data.go
Comment (efiShtain) [reply internal/backend/sessions/sessiondata/data.go:63]: it just make the system more complicated for stuff we don't really need, nor there are requirements for it
I think error it better here

File: internal/backend/sessions/sessiondata/data.go
Comment (efiShtain) [reply internal/backend/sessions/sessiondata/data.go:56]: ok, 
thought this is an external API

File: internal/backend/db/dbgorm/projects.go
Comment (efiShtain) [reply internal/backend/db/dbgorm/projects.go:56]:
```
// enforce foreign keys constrains while soft-deleting - should be no active deployments
```

the comment state it should check active deployments 

so which is wrong? code or comment ? 
I guess the comment since the context is delete env

**Review by efiShtain — APPROVED** [review]
some stuff, not a blocker for such a big change

File: cmd/ak/common/build.go
Comment (itayd) [reply cmd/ak/common/build.go]: that's what i did now...

File: cmd/ak/cmd/projects/deploy.go
Comment (itayd) [reply cmd/ak/cmd/projects/deploy.go:50]: the api lets you build arbitrary stuff. we'll limit it with rbac later.

File: internal/backend/sessions/sessiondata/data.go
Comment (itayd) [reply internal/backend/sessions/sessiondata/data.go:63]: i think it is useful, not just for testing. will limit it with rbac later.

File: proto/autokitteh/dispatcher/v1/svc.proto
Comment (itayd) [reply proto/autokitteh/dispatcher/v1/svc.proto:29]: you can have "testing" deployments. these are deployments that receive events that are specifically sent to them.

**Pull request merged**
